records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,178 @@
+    id8-->id9((X))
+
+```
+
+### Use magic links
+The industry standard for magic links is to provide passwordless login for users. They are a token based authentication strategy that act as a replacement for password input.
+
+In the context of Notify, magic links can be adapted to serve as a simpler second step in the two-factor authentication process. After inputting their username and password, the user checks their email and clicks the magic link to complete the login process. With this approach we can make the login process simpler and more accessible by reducing the number of steps required by the user to log in, thus making the process less obstructive and feel more natural.
+
+
+```mermaid
+
+flowchart LR
+	subgraph Notify Tab
+        id1(Enter username and password)
+        id4(User is logged into Notify)
+    end
+    subgraph Email Inbox Tab
+        id2(Find the magic link)-->id3(Click the link)
+    end
+    id1-->id2
+    id3-->id4
+    id4-->id5((X))
+
+```
+
+## Security Considerations
+While the addition of magic links provides users with a less cumbersome, more accessible and convenient login experience there are a number of security implications that we identified during a [threat modeling exercise](https://docs.google.com/document/d/1sIdrJ_dlkzwQoEVss0SMXjhdrLPYUenFlf3Xm8-amUc) that we should mitigate.
+
+### Session cookie leaks via Cross-Site-Scripting
+Once a user inputs their username and password, a partially authenticated session cookie is created that could, theoretically, be used by an attacker to gain unauthorized access to Notify. For this to succeed the attacker needs:
+
+1. Access to a user's partially authenticated cookie.
+2. Access to  the user's email inbox to obtain their magic link.
+   * Man-in-the-middle attack to obtain the user's magic link.
+
+
+#### Mitigations
+* CSRF protections buit into Flask, and which are already enabled in our application, configure the `Set-Cookie` header with the `HttpOnly` and `Secure` options, preventing Javascript from reading the contents of the session cookie, and limits the cookie(s) to HTTPS traffic only
+
+* The `Strict-Transport-Security` header is enabled site wide, converting all requests to HTTPS, thus mitigating man-in-the-middle attacks.
+
+* Introduce a process for checking that the `Origin` header matches the target origin when a magic link is followed. As our application is distributed and lives behind a load balancer, making each node aware of it's target origin could prove to be non-trivial task and/or have unintended consequences.
+
+See the [Fallback Mitigation section](#Fallback-Mitigation---An-over-engineered-"last-resort"-solution)
+
+### Email Phishing
+An attacker could tamper with the 2FA email we send to the user and forward them to a phishing site that prompts them to re-input their username and password. If the user is not aware of this type of risk they could unintentionally provide an attacker with their log in credentials.
```

Comment:
IMHO, a more likely scenario (and probably easier to achieve) is to create an e-mail which looks like Notify's 2FA email, but with a malicious link (e.g. reflected XSS). The following measures could mitigate this:

- ensure Notify is not vulnerable to XSS (e.g. via automated vulnerability scanning & patching).
- educating end user on how to identify legitimate (e.g. verifying the from field in e-mail header) e-mails from Notify.